lib/pharos/phase_manager.rb

```diff
@@ -30,12 +30,23 @@ def initialize(**options)
     # @param phases [Array<Pharos::Phases::Base>]
     # @return [Array<...>]
     def run_parallel(phases, &block)
-      threads = phases.map { |phase|
+      threads = phases.map do |phase|
         Thread.new do
           Thread.current.report_on_exception = false
           Retry.perform(yield_object: phase, logger: logger, exceptions: RETRY_ERRORS, &block)
         end
-      }
+      end
+
+      until threads.none?(&:alive?)
```

Comment:
This is where the sauce is.

> `abort_on_exception` → true or false
> When set to true, if any thread is aborted by an exception, the raised exception will be re-raised in the main thread.

> `report_on_exception → true or false`
> All threads created when this flag is true will report a message on $stderr if an exception kills the thread.
> This is done to catch errors in threads early. In some cases, you might not want this output. There are multiple ways to avoid the extra output:
> * If the exception is not intended, the best is to fix the cause of the exception so it does not happen anymore.
> * If the exception is intended, it might be better to rescue it closer to where it is raised rather then let it kill the Thread.
> * **If it is guaranteed the Thread will be joined with #join or #value, then it is safe to disable this report** with Thread.current.report_on_exception = false when starting the Thread. However, this might handle the exception much later, or not at all if the Thread is never joined due to the parent thread being blocked, etc

The problem with `abort_on_exception=true` is that the main thread may be anywhere when that happens, which messes up the exception handling and it will miss `Pharos::Command`'s `rescue`s. Relying on `thread.value` to raise, it's possible to rescue reliably exactly where expected.

If the other threads are not killed, they may keep running and outputting until the main thread exits.